SparseGrids/tsgLinearSolvers.cpp

```diff
@@ -237,6 +256,97 @@ void TasmanianTridiagonalSolver::decompose(int n, std::vector<double> &d, std::v
     }
 }
 
+void TasmanianTridiagonalSolver::decompose2(std::vector<double> &diag, std::vector<double> &off_diag, const double mu0,
+                                            std::vector<double> &nodes, std::vector<double> &weights) {
+
+    // Ensure compatibility with the ALGOL implementation.
+    // NOTE: diag and off_diag are 1-indexed, while nodes and weights are 0-indexed after this step.
+    size_t n = diag.size();
+    if (off_diag.size() != n-1) {
+        throw std::invalid_argument("ERROR: off_diag.size() must be equal to diag.size()-1");
+    }
+    nodes.resize(n);
+    weights.resize(n);
+    diag.insert(diag.begin(), 0.0);
+    off_diag.insert(off_diag.begin(), 0.0);
+
+    // SETUP block from ALGOL code.
+    // ALGOL COMMENT: Find the maximum row sum norm and initialize weights.
+    off_diag[0] = 0.0;
+    double norm = 0.0;
+    for (size_t i=1; i<=n-1; i++) {
+        norm = std::max(norm, std::fabs(off_diag[i-1]) + std::fabs(diag[i]) + std::fabs(off_diag[i-1]));
+        weights[i-1] = 0.0;
+    }
+    norm = std::max(norm, std::fabs(diag[n]) + std::fabs(off_diag[n-1]));
+    weights[n-1] = 0.0;
+    weights[0] = 1.0; // Fix the bug in the ALGOL code.
+    double eps = norm * std::numeric_limits<double>::epsilon();
+    size_t m = n;
+    double lambda{norm}, lambda1{norm}, lambda2{norm}, rho{norm};
+
+    // INSPECT block from ALGOL code.
+    // ALGOL COMMENT: Look for convergence of lower diagonal element.
+    while (m > 0) {
+        if (std::fabs(off_diag[m-1]) <= eps) {
+            nodes[m-1] = diag[m];
+            weights[m-1] = mu0 * weights[m-1] * weights[m-1];
+            rho = lambda1 < lambda1 ? lambda1 : lambda2;
+            m = m-1;
```

Comment:
Use `m--;`

It's Ok to blindly copy-paste code in the first iteration, but then fix the style to match the language that you're using.

Reply:
Done.